File: hw/hw03/hw03.swift
func num_eights(_ pos: Int) -> Int {
    if pos == 0 {
        return 0
    }
    if pos % 10 == 8 {
        return 1 + num_eights(pos / 10)
    } else {
        return num_eights(pos / 10)
    }
}

print("DEBUG:", num_eights(88888888))

/**
 iteration 
 
 */

func pingpong(_ n: Int) -> Int {
//    var idx = 1
//    var sign = 1
//    var res = 0
//    while idx <= n {
//        res += sign
//        sign *= changeSign(idx)
//        idx += 1
//    }
//    return res
    return pingpong_recursive(final: n, index: 1, sign: 1)
}

func changeSign(_ idx: Int) -> Int {
    if idx % 8 == 0 || has_eight(idx) {
        return -1
    }
    return 1
}

func has_eight(_ k: Int) -> Bool {
    if k % 10 == 8 {
        return true
    } else if k <= 0 {
        return false
    }
    return has_eight(k / 10)
}
//print("DEBUG:", has_eight(70))
//
//print("DEBUG:", pingpong(30))

func pingpong_recursive(final: Int, index: Int, sign: Int) -> Int {
    if index <= final {
        var signChanged = sign * changeSign(index)
        var idx = index + 1
        print("DEBUG: ", index, " ", sign)
        return sign + pingpong_recursive(final: final, index: idx, sign: signChanged)
    } else {
        return 0
    }
}

print("DEBUG:", pingpong(18))

File: hw/hw03/hw03.py
HW_SOURCE_FILE = __file__


def num_eights(pos):
    """Returns the number of times 8 appears as a digit of pos.

    >>> num_eights(3)
    0
    >>> num_eights(8)
    1
    >>> num_eights(88888888)
    8
    >>> num_eights(2638)
    1
    >>> num_eights(86380)
    2
    >>> num_eights(12345)
    0
    >>> from construct_check import check
    >>> # ban all assignment statements
    >>> check(HW_SOURCE_FILE, 'num_eights',
    ...       ['Assign', 'AnnAssign', 'AugAssign', 'NamedExpr'])
    True
    """
    "*** YOUR CODE HERE ***"
    if pos == 0:
        return 0
    if pos % 10 == 8:
        return 1 + num_eights(pos // 10)
    else: return num_eights(pos // 10)


def pingpong(n):
    """Return the nth element of the ping-pong sequence.

    >>> pingpong(8)
    8
    >>> pingpong(10)
    6
    >>> pingpong(15)
    1
    >>> pingpong(21)
    -1
    >>> pingpong(22)
    -2
    >>> pingpong(30)
    -2
    >>> pingpong(68)
    0
    >>> pingpong(69)
    -1
    >>> pingpong(80)
    0
    >>> pingpong(81)
    1
    >>> pingpong(82)
    0
    >>> pingpong(100)
    -6
    >>> from construct_check import check
    >>> # ban assignment statements
    >>> check(HW_SOURCE_FILE, 'pingpong',
    ...       ['Assign', 'AnnAssign', 'AugAssign', 'NamedExpr'])
    True
    """
    "*** YOUR CODE HERE ***"
    def changeSign(index):
        if index % 8 == 0 or num_eights(index) > 0:
            return -1
        else: return 1

    def helper(final, index, sign):
        if index <= final:
            return sign + helper(final, index + 1, sign * changeSign(index))
        else: return 0
    
    return helper(n, 1, 1)


def get_larger_coin(coin):
    """Returns the next larger coin in order.
    >>> get_larger_coin(1)
    5
    >>> get_larger_coin(5)
    10
    >>> get_larger_coin(10)
    25
    >>> get_larger_coin(2) # Other values return None
    """
    if coin == 1:
        return 5
    elif coin == 5:
        return 10
    elif coin == 10:
        return 25


def get_smaller_coin(coin):
    """Returns the next smaller coin in order.
    >>> get_smaller_coin(25)
    10
    >>> get_smaller_coin(10)
    5
    >>> get_smaller_coin(5)
    1
    >>> get_smaller_coin(2) # Other values return None
    """
    if coin == 25:
        return 10
    elif coin == 10:
        return 5
    elif coin == 5:
        return 1


def count_coins(change):
    """Return the number of ways to make change using coins of value of 1, 5, 10, 25.
    >>> count_coins(15)
    6
    >>> count_coins(10)
    4
    >>> count_coins(20)
    9
    >>> count_coins(100) # How many ways to make change for a dollar?
    242
    >>> count_coins(200)
    1463
    >>> from construct_check import check
    >>> # ban iteration
    >>> check(HW_SOURCE_FILE, 'count_coins', ['While', 'For'])
    True
    """
    "*** YOUR CODE HERE ***"
    def count_partitions(to_sum_change, partition):
        # if max_of_partition == "":
        #     return 1
        # elif to_sum_change == "":
        #     return 1
        # elif to_sum_change 

        def nearest_coin(partition):
            if partition == 1:
                return 1
            elif get_smaller_coin(partition) == None:
                return nearest_coin(partition - 1)
            else:
                return partition

        if partition == 1:
            return 1
        elif to_sum_change == 0:
            return 1
        elif to_sum_change < 0:
            return 0
        
        if get_smaller_coin(partition) == None:
            partition = nearest_coin(partition)
        with_max_partition = count_partitions(to_sum_change - partition, partition)
        without_max_partition = count_partitions(to_sum_change, get_smaller_coin(partition))

        return with_max_partition + without_max_partition

    return count_partitions(change, change)
